// forta-agent-tools/lib/utils/victim-identification/helpers/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MKR_TOKEN_ABI = exports.TOKEN_ABI = exports.WRAPPED_NATIVE_TOKEN_EVENTS = exports.ERC20_TRANSFER_EVENT = exports.ZERO = exports.SUBGRAPH_URL = exports.PREPARATION_BOT = exports.MAX_USD_VALUE = void 0;
var forta_agent_1 = require("forta-agent");
exports.MAX_USD_VALUE = 500000;
exports.PREPARATION_BOT = ["0x0b241032ca430d9c02eaa6a52d217bbff046f0d1b3f3d2aa928e42a97150ec91"]; // Malicious Contract Creation ML Bot v2
exports.SUBGRAPH_URL = "https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v3";
exports.ZERO = forta_agent_1.ethers.constants.Zero;
exports.ERC20_TRANSFER_EVENT = "event Transfer(address indexed from, address indexed to, uint256 value)";
exports.WRAPPED_NATIVE_TOKEN_EVENTS = [
    "event Deposit(address indexed to, uint256 value)",
    "event Withdrawal(address indexed from, uint256 value)",
];
exports.TOKEN_ABI = [
    "function balanceOf(address account) external view returns (uint256)",
    "function symbol() external view returns (string)",
    "function name() public view returns (string)",
    "function decimals() external view returns (uint8)",
    "function totalSupply() external view returns (uint256)",
];
exports.MKR_TOKEN_ABI = ["function symbol() external view returns (bytes32)"];
